assets/sip-big_vector/big_vector.move

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) Typus Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module typus_framework::big_vector {
+    use std::vector;
+
+    use sui::object::{Self, UID};
+    use sui::tx_context::TxContext;
+    use sui::dynamic_field;
+
+    // ======== Errors ========
+
+    const E_NOT_EMPTY: u64 = 0;
+
+    // ======== Structs ========
+
+    struct BigVector<phantom Element> has key, store {
+        id: UID,
+        slice_count: u64,
+        slice_size: u64,
+        length: u64,
+    }
+
+    // ======== Functions ========
+
+    /// create BigVector
+    public fun new<Element: store>(slice_size: u64, ctx: &mut TxContext): BigVector<Element> {
+        let id = object::new(ctx);
+        let slice_count = 1;
+        dynamic_field::add(&mut id, slice_count, vector::empty<Element>());
+        BigVector<Element> {
+            id,
+            slice_count,
+            slice_size,
+            length: 0,
+        }
+    }
+
+    /// return the slice_count of the BigVector
+    public fun slice_count<Element: store>(bv: &BigVector<Element>): u64 {
+        bv.slice_count
+    }
+
+    /// return the max size of a slice in the BigVector
+    public fun slice_size<Element: store>(bv: &BigVector<Element>): u64 {
+        bv.slice_size
+    }
+
+    /// return the size of the BigVector
+    public fun length<Element: store>(bv: &BigVector<Element>): u64 {
+        bv.length
+    }
+
+    /// return the slice_id related to the index i
+    public fun slice_id<Element: store>(bv: &BigVector<Element>, i: u64): u64 {
+        (i / bv.slice_size) + 1
+    }
+
+    /// return true if the BigVector is empty
+    public fun is_empty<Element: store>(bv: &BigVector<Element>): bool {
+        bv.length == 0
+    }
+
+    /// push a new element at the end of the BigVector
+    public fun push_back<Element: store>(bv: &mut BigVector<Element>, element: Element) {
+        if (length(bv) / bv.slice_size == bv.slice_count) {
+            bv.slice_count = bv.slice_count + 1;
+            let new_slice = vector::singleton(element);
+            dynamic_field::add(&mut bv.id, bv.slice_count, new_slice);
+        }
+        else {
+            let slice = dynamic_field::borrow_mut(&mut bv.id, bv.slice_count);
+            vector::push_back(slice, element);
+        };
+        bv.length = bv.length + 1;
+    }
+
+    /// pop an element from the end of the BigVector
+    public fun pop_back<Element: store>(bv: &mut BigVector<Element>): Element {
+        let slice = dynamic_field::borrow_mut(&mut bv.id, bv.slice_count);
+        let element = vector::pop_back(slice);
+        trim_slice(bv);
+        bv.length = bv.length - 1;
+
+        element
+    }
```

Comment:
Looking at this API, I notice that most of the mutating operations work on the end of the vector.  There's a potential optimisation here where you store the last slice in the `BigVector` inline:

```rust
struct BigVector<Element> has key, store {
    id: UID,
    slice_size: u32,
    slice_count: u64,
    tail: vector<Element>,
}
```

This means that pushes, pops and swap removes don't need to load a dynamic field to access the last slice.  `remove` is a problem however.

Unfortunately, today, we don't have a way to take `tail` and replace it with an empty vector (but taking its contents to reuse elsewhere), but I think that should be feasible/safe to add? (cc @dariorussi) That would make the operation of spilling the tail into a new dynamic field much cheaper.

Reply:
certainly an interesting usage for the replace operator (@tnowacki)

Reply:
However, if we wrap the vector to the BigVector struct, the object size of this struct grows with the vector. Although most of the mutating operations work on the end of the vector, it accesses only a single dynamic_field and I think the cost is reasonable.